### PR TITLE
En 5594 add disable prune flag in config

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**

### CODE REVIEW

File: process/block/shardblock.go [hunk 1]
```diff
@@ -813,6 +813,10 @@ func (sp *shardProcessor) CommitBlock(
 	}
 
 	for i := range finalHeaders {
```
Comment (sasurobert) [inline process/block/shardblock.go]: blockProcessor should not know about this. we could delete everything before final nonce. Add TODO. Move for i := range finalHeaders to a new function.

File: cmd/node/config/config.toml [hunk 2]
```diff
@@ -344,5 +344,6 @@
    TimeoutMilliseconds = 100
    Version = 0  # Setting 0 means 'use default value'
 
-[StateCheckpointConfig]
+[StateTrieConfig]
     RoundsModulus = 100
+    PruningEnabled = false
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply cmd/node/config/config.toml]: Done.

File: data/trie/patriciaMerkleTrie.go [hunk 3]
```diff
@@ -74,7 +76,7 @@ func (tr *patriciaMerkleTrie) Get(key []byte) ([]byte, error) {
 
 	val, err := tr.root.tryGet(hexKey, tr.trieStorage.Database())
 	if err != nil {
-		log.Trace("trie get", "error", key)
+		err = fmt.Errorf("trie get error: %w, for key %v", err, hex.EncodeToString(key))
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/trie/patriciaMerkleTrie.go]: The error is logged at a higher level. I need to know that the error comes from "trie get" function and not from other function that can return the same err.

File: data/trie/patriciaMerkleTrie.go [hunk 4]
```diff
@@ -326,7 +328,7 @@ func (tr *patriciaMerkleTrie) Recreate(root []byte) (data.Trie, error) {
 
 	newTr, err := tr.recreateFromDb(root)
 	if err != nil {
-		log.Debug("trie recreate", "error", root)
+		err = fmt.Errorf("trie recreate error: %w, for root %v", err, core.ToB64(root))
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/trie/patriciaMerkleTrie.go]: The error is logged at a higher level. I need to know that the error comes from "trie recreate" function and not from other function that can return the same err.

File: integrationTests/state/stateTrie_test.go [hunk 5]
```diff
@@ -1278,8 +1278,9 @@ func TestRollbackBlockAndCheckThatPruningIsCancelled(t *testing.T) {
 	assert.Equal(t, uint64(3), nodes[0].BlockChain.GetCurrentBlockHeader().GetNonce())
 	assert.Equal(t, uint64(4), nodes[1].BlockChain.GetCurrentBlockHeader().GetNonce())
 
+	expectedErr := fmt.Errorf("trie recreate error: %w, for root %v", trie.ErrHashNotFound, core.ToB64(rootHashOfRollbackedBlock))
 	err = shardNode.AccntState.RecreateTrie(rootHashOfRollbackedBlock)
-	assert.Equal(t, trie.ErrHashNotFound, err)
+	assert.Equal(t, expectedErr, err)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply integrationTests/state/stateTrie_test.go]: Done.

File: integrationTests/state/stateTrie_test.go [hunk 6]
```diff
@@ -1405,8 +1406,9 @@ func TestTriePruningWhenBlockIsFinal(t *testing.T) {
 	assert.Equal(t, uint64(7), nodes[0].BlockChain.GetCurrentBlockHeader().GetNonce())
 	assert.Equal(t, uint64(7), nodes[1].BlockChain.GetCurrentBlockHeader().GetNonce())
 
+	expectedErr := fmt.Errorf("trie recreate error: %w, for root %v", trie.ErrHashNotFound, core.ToB64(rootHashOfFirstBlock))
 	err := shardNode.AccntState.RecreateTrie(rootHashOfFirstBlock)
-	assert.Equal(t, trie.ErrHashNotFound, err)
+	assert.Equal(t, expectedErr, err)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply integrationTests/state/stateTrie_test.go]: Done.

File: process/sync/baseSync.go [hunk 7]
```diff
@@ -704,12 +704,15 @@ func (boot *baseBootstrap) rollBackOneBlock(
 			return err
 		}
 
-		boot.accounts.CancelPrune(prevHeader.GetRootHash())
-
-		if !bytes.Equal(currHeader.GetRootHash(), prevHeader.GetRootHash()) {
-			errNotCritical := boot.accounts.PruneTrie(currHeader.GetRootHash())
-			if errNotCritical != nil {
-				log.Debug(errNotCritical.Error())
+		if boot.accounts.IsPruningEnabled() {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply process/sync/baseSync.go]: Added a TODO to check this code when the pruning mechanism will be changed.

File: process/block/shardblock.go [hunk 8]
```diff
@@ -813,6 +813,10 @@ func (sp *shardProcessor) CommitBlock(
...
+		if !sp.accounts.IsPruningEnabled() {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply process/block/shardblock.go]: Added a TODO.

File: process/block/shardblock.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply process/block/shardblock.go]: Added TODO.

File: data/trie/factory/trieFactoryArgs.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+	snapshotDbCfg          config.DBConfig
+	marshalizer            marshal.Marshalizer
+	hasher                 hashing.Hasher
+	pathManager            storage.PathManagerHandler
+	shardId                string
+	pruningEnabled         bool
+}
+
+// NewTrieFactoryArgs returns a struct that holds all arguments needed for creating a trie factory
+func NewTrieFactoryArgs(
```
Comment (sasurobert) [inline data/trie/factory/trieFactoryArgs.go]: this can be done without factory. Args can be public.

File: data/trie/factory/trieCreator.go [hunk 10]
```diff
@@ -22,71 +22,71 @@ type trieCreator struct {
 
 // NewTrieFactory creates a new trie factory
 func NewTrieFactory(
-	args *trieFactoryArgs,
+	args *TrieFactoryArgs,
```
Comment (sasurobert) [inline data/trie/factory/trieCreator.go]: no need for pointer here.

**Pull request merged**